commons/src/lexer/token/iterator/mod.rs

```diff
@@ -0,0 +1,1284 @@
+            prefix_start,
+        }
+    }
+
+    /// Nests this iterator, by creating a new iterator that has this iterator set as parent.
+    /// Pushes the new iterator to a new scope root, and only runs the given matching functions in the new scope.
+    /// Parent iterators do not get updated to the new scope, so matching functions for scoped parent iterators are still executed.
+    ///
+    /// # Arguments
+    ///
+    /// * `prefix_match` ... Optional matching function used to strip prefix on new lines
+    /// * `end_match` ... Optional matching function used to indicate the end of the created iterator
+    pub fn new_scope_root(
+        self,
+        prefix_match: Option<IteratorPrefixFn>,
+        end_match: Option<IteratorEndFn>,
+    ) -> TokenIterator<'slice, 'input> {
+        let start_index = self.index();
+
+        TokenIterator {
+            parent: TokenIteratorKind::ScopedRoot(TokenIteratorScopedRoot::from(self)),
+            scope: 0,
+            scoped: false,
+            skip_end_until_idx: 0,
+            start_index,
+            match_index: 0,
+            prefix_match,
+            end_match,
+            iter_end: false,
+            prefix_mismatch: false,
+            next_matching: false,
+            peek_matching: false,
+            prefix_start: None,
+        }
+    }
+
+    /// Returns `true` if this iterator is nested.
+    pub fn is_nested(&self) -> bool {
+        matches!(
+            self.parent,
+            TokenIteratorKind::Nested(_) | TokenIteratorKind::ScopedRoot(_)
+        )
+    }
+
+    /// Returns `true` if this iterator created a new scope.
+    pub fn is_scoped(&self) -> bool {
+        self.scoped
+    }
+
+    /// Returns the parent iterator if this iterator is nested.
+    pub fn unfold(self) -> Self {
+        match self.parent {
+            TokenIteratorKind::Nested(mut parent) => {
+                parent.set_curr_scope(parent.scope);
+                *parent
+            }
+            TokenIteratorKind::Root(_) => {
+                debug_assert!(false, "Tried to unfold non-nested iterator.");
+                self
+            }
+            TokenIteratorKind::ScopedRoot(scoped_root) => *scoped_root.token_iter,
+        }
+    }
+
+    /// Tries to skip tokens until one of the end functions signals the end.
+    ///
+    /// **Note:** This function might not reach the iterator end.
+    ///
+    /// If no tokens are left, or no given line prefix is matched, the iterator may stop before an end is reached.
+    /// Use [`Self::end_reached()`] to check if the end was actually reached.
+    pub fn skip_to_end(mut self) -> Self {
+        let _last_token = self.by_ref().last();
+
+        self
+    }
+
+    pub fn peek_nth(&mut self, n: usize) -> Option<&'slice Token<'input>> {
+        let mut token = self.peeking_next(|_| true);
+
+        for _ in 0..n {
+            token = self.peeking_next(|_| true);
+            token?;
+        }
+
+        token
```

Comment:
I think this can be written in a more elegant way:

```suggestion
        std::iter::repeat_with(|| self.peeking_next(|_| true)).nth(n).flatten()
```

Reply:
Thanks :)
Not sure if I'd call it easier to read, but definitely looks more elegant :D